core/src/types/operator/operator_futures.rs

```diff
@@ -0,0 +1,126 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Futures provides the futures generated by [`Operator`]
+//!
+//! By using futures, users can add more options for operation.
+
+use std::mem;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use futures::future::BoxFuture;
+use futures::Future;
+use futures::FutureExt;
+
+use crate::ops::*;
+use crate::raw::*;
+use crate::*;
+
+/// OperatorFuture is the future generated by [`Operator`].
+///
+/// The future will consume all the input to generate a future.
+pub(crate) enum OperatorFuture<T, F> {
+    /// Idle state, waiting for the future to be polled
+    Idle(
+        /// The accessor to the underlying object storage
+        FusedAccessor,
+        /// The path of string
+        String,
+        /// The input args
+        T,
+        /// The function which will move all the args and return a static future
+        fn(FusedAccessor, String, T) -> BoxFuture<'static, Result<F>>,
+    ),
+    /// Polling state, waiting for the future to be ready
+    Poll(BoxFuture<'static, Result<F>>),
+    /// Done state, the future has been polled and completed or something is
+    /// broken during state switch.
+    Done,
+}
+
+impl<T, F> OperatorFuture<T, F> {
+    pub fn new(
+        inner: FusedAccessor,
+        path: String,
+        args: T,
+        f: fn(FusedAccessor, String, T) -> BoxFuture<'static, Result<F>>,
+    ) -> Self {
+        OperatorFuture::Idle(inner, path, args, f)
+    }
+
+    fn map_args(self, f: impl FnOnce(T) -> T) -> Self {
+        match self {
+            OperatorFuture::Idle(inner, path, args, func) => {
+                OperatorFuture::Idle(inner, path, f(args), func)
+            }
+            _ => unreachable!("future has been polled and should not be changed again"),
+        }
+    }
+}
+
+impl<T, F> Future for OperatorFuture<T, F>
+where
+    T: Unpin,
+    F: Unpin,
+{
+    type Output = Result<F>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        *self = match mem::replace(self.as_mut().get_mut(), OperatorFuture::Done) {
```

Comment:
Is this the only way here?

Reply:
> Is this the only way here?

The best code I can(

Reply:
> This doesn't look like a `Done` after replacing either.

I am guessing you didn't understand code here correctly, let me add more comments here.